### ISSUE TITLE
Setting mapgen objects

### ISSUE THREAD

**Issue by basxto** [minetest/minetest#13741]
## Problem

`minetest.get_mapgen_object()` allows to get mapgen objects, but there is no way of setting them. Many mods don’t check correctly if they get `nil` returned and just don’t work even worse crash with lua mapgens.

## Solutions

Introduction of a function for setting mapgen objects for the current mapgen thread.

## Additional context

Some lua map/biome generators have that data for a current chunk, but they have no way of sharing it with other mods.

**Comment by sfan5** [comment]
Which of the types that `get_mapgen_object` can return are you talking about?

**Comment by basxto** [comment]
`heightmap`, `biomemap`, `heatmap` and `humiditymap`

**Comment by fluxionary** [comment]
> Many mods don’t check correctly if they get nil returned and just don’t work even worse crash with lua mapgens.

that's on the mods, it's well documented that it can return `nil`.

> Some lua map/biome generators have that data for a current chunk, but they have no way of sharing it with other mods.

is that the intended use case, to allow code that's not inside an `on_generate()` callback to have access to those mapgen objects? i have to admit, i'd *love* to have access to the heightmap, as well as the after-generation biomemap (for the valleys mapgen in particular). this would just require persisting that data somehow (w/ the mapblock?), and would still be missing for any "old" areas of the map generated before the implementation.

i'm not sure that a general "set" mechanism makes sense. i wouldn't think that modifying those values after generation would be useful for much of anything, am i wrong?

**Comment by basxto** [comment]
No, this is about the `on_generate()` context. As far as I understand lua mapgens, they populate the chunk through `on_generate()` after singlenode mapgen is done. That’s at least how `mapgen_rivers` and `lvm_example` work.

The goal is to relay information to `on_generate()` callbacks that run after it in the same mapgen thread. And to discard it after generation is finished.

> that's on the mods, it's well documented that it can return nil.

Yes. I have seen both … crashing with singlenode (dfcaverns) and disabling parts of themselves for singlenode (mineclone2).

The other way would be that mods have to detect mapgen/biomegen and register with a custom `on_generate()` that gives those maps as parameters to the callback function.

**Comment by sfan5** [comment]
Note that this involves a bit more than just providing the mapgen objects since `minetest.get_heat` and `minetest.get_humidity` exist too.
